lib/Interpreter/CXCppInterOp.cpp

```diff
@@ -0,0 +1,627 @@
+}
+
+namespace Cpp {
+bool InsertOrReplaceJitSymbol(compat::Interpreter& I,
+                              const char* linker_mangled_name,
+                              uint64_t address);
+} // namespace Cpp
+
+bool clang_Interpreter_insertOrReplaceJitSymbol(CXInterpreter I,
+                                                const char* linker_mangled_name,
+                                                uint64_t address) {
+  return Cpp::InsertOrReplaceJitSymbol(*getInterpreter(I), linker_mangled_name,
+                                       address);
+}
+
+static inline clang::QualType getType(const CXQualType& Ty) {
+  return clang::QualType::getFromOpaquePtr(Ty.data[0]);
+}
+
+static inline CXInterpreterImpl* getNewTU(const CXQualType& Ty) {
+  return static_cast<CXInterpreterImpl*>(Ty.data[1]);
+}
+
+static inline compat::Interpreter* getInterpreter(const CXQualType& Ty) {
+  return getInterpreter(static_cast<const CXInterpreterImpl*>(Ty.data[1]));
+}
+
+CXString clang_getTypeAsString(CXQualType type) {
+  const clang::QualType QT = getType(type);
+  const auto& C = getInterpreter(type)->getSema().getASTContext();
+  clang::PrintingPolicy Policy = C.getPrintingPolicy();
+  Policy.Bool = true;               // Print bool instead of _Bool.
+  Policy.SuppressTagKeyword = true; // Do not print `class std::string`.
+  return makeCXString(compat::FixTypeName(QT.getAsString(Policy)));
+}
+
+CXQualType clang_getComplexType(CXQualType eltype) {
+  const auto& C = getInterpreter(eltype)->getSema().getASTContext();
+  return clang::cxscope::MakeCXQualType(C.getComplexType(getType(eltype)),
+                                        getNewTU(eltype));
+}
+
+static inline bool isNull(const CXScope& S) { return !S.data[0]; }
+
+static inline clang::Decl* getDecl(const CXScope& S) {
+  return const_cast<clang::Decl*>(static_cast<const clang::Decl*>(S.data[0]));
+}
+
+static inline const CXInterpreterImpl* getNewTU(const CXScope& S) {
+  return static_cast<const CXInterpreterImpl*>(S.data[2]);
+}
+
+static inline CXCursorKind kind(const CXScope& S) { return S.kind; }
+
+static inline compat::Interpreter* getInterpreter(const CXScope& S) {
+  return getInterpreter(static_cast<const CXInterpreterImpl*>(S.data[2]));
+}
+
+void clang_scope_dump(CXScope S) { getDecl(S)->dump(); }
+
+bool clang_hasDefaultConstructor(CXScope S) {
+  auto* D = getDecl(S);
+
+  if (const auto* CXXRD = llvm::dyn_cast_or_null<clang::CXXRecordDecl>(D))
+    return CXXRD->hasDefaultConstructor();
+
+  return false;
+}
+
+CXScope clang_getDefaultConstructor(CXScope S) {
+  if (!clang_hasDefaultConstructor(S))
+    return clang::cxscope::MakeCXScope(nullptr, getNewTU(S));
+
+  auto* CXXRD = llvm::dyn_cast_or_null<clang::CXXRecordDecl>(getDecl(S));
+  if (!CXXRD)
+    return clang::cxscope::MakeCXScope(nullptr, getNewTU(S));
+
+  const auto* Res =
+      getInterpreter(S)->getSema().LookupDefaultConstructor(CXXRD);
+  return clang::cxscope::MakeCXScope(Res, getNewTU(S));
```

Comment:
Why can't we forward to `Cpp::GetDefaultConstructor`? I see this trend in several interfaces. Is there some limitation on the C++ side?

Reply:
There are no particular reasons. I forwarded this one to the C++ API. As for other interpreter-related APIs, the C++ implementation is also a thin wrapper, so I copied the implementation on the C API side.